### PR TITLE
Fix gradient checkpointing bugs in freezing part of models (requires_grad=False)

### PR DESCRIPTION
**Pull request by 7eu7d7**
Using ```torch.utils.checkpoint.checkpoint``` directly will cause the parameters in the checkpoint section to not be learned when part of the model parameters are freezed. As these discussions state:

https://discuss.pytorch.org/t/use-of-torch-utils-checkpoint-checkpoint-causes-simple-model-to-diverge/116271

https://discuss.pytorch.org/t/checkpoint-with-no-grad-requiring-inputs-problem/19117/19

In pytroch versions larger than 1.11.0, the ```use_reentrant=False`` can be added to fix this bug.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @7eu7d7,

Could you give a bit more context for this PR?

**Comment by 7eu7d7** [comment]
> Hey @7eu7d7,
> 
> Could you give a bit more context for this PR?

When adding adapters (like lora) to the model or training only partial layers of the model (The rest of the parameters do not requires gradients), enable gradient checkpointing can cause some parameters that require gradients cannot be learned. This bug is also existed in the ```transformers```.

File: src/diffusers/models/unet_2d_blocks.py [hunk 1]
```diff
@@ -856,13 +856,23 @@ def custom_forward(*inputs):
...
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
-                hidden_states = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(attn, return_dict=False),
-                    hidden_states,
-                    encoder_hidden_states,
-                    cross_attention_kwargs,
-                )[0]
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:859]:
```suggestion
                if is_torch_version("<", "2.0.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 2]
```diff
@@ -1340,7 +1353,10 @@ def custom_forward(*inputs):
 
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1356]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 3]
```diff
@@ -1537,7 +1553,10 @@ def custom_forward(*inputs):
...
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1556]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 4]
```diff
@@ -1848,13 +1878,23 @@ def custom_forward(*inputs):
...
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
-                hidden_states = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(attn, return_dict=False),
-                    hidden_states,
-                    encoder_hidden_states,
-                    cross_attention_kwargs,
-                )[0]
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1881]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 5]
```diff
@@ -1934,7 +1974,10 @@ def custom_forward(*inputs):
 
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:1977]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 6]
```diff
@@ -2351,7 +2394,10 @@ def custom_forward(*inputs):
...
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2397]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 7]
```diff
@@ -2556,7 +2602,10 @@ def custom_forward(*inputs):
...
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2605]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/unet_2d_blocks.py [hunk 8]
```diff
@@ -2677,14 +2726,25 @@ def custom_forward(*inputs):
...
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
-                hidden_states = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(attn, return_dict=False),
-                    hidden_states,
-                    encoder_hidden_states,
-                    attention_mask,
-                    cross_attention_kwargs,
-                )[0]
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_blocks.py:2729]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/models/vae.py [hunk 9]
```diff
@@ -221,13 +225,22 @@ def custom_forward(*inputs):
 
                 return custom_forward
 
-            # middle
-            sample = torch.utils.checkpoint.checkpoint(create_custom_forward(self.mid_block), sample)
-            sample = sample.to(upscale_dtype)
+            if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/models/vae.py:228]:
```suggestion
            if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 10]
```diff
@@ -1077,7 +1077,10 @@ def custom_forward(*inputs):
 
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:1080]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 11]
```diff
@@ -1198,13 +1201,23 @@ def custom_forward(*inputs):
...
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
-                hidden_states = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(attn, return_dict=False),
-                    hidden_states,
-                    encoder_hidden_states,
-                    cross_attention_kwargs,
-                )[0]
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:1204]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 12]
```diff
@@ -1289,7 +1302,10 @@ def custom_forward(*inputs):
 
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:1305]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

File: src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py [hunk 13]
```diff
@@ -1412,13 +1428,23 @@ def custom_forward(*inputs):
...
                     return custom_forward
 
-                hidden_states = torch.utils.checkpoint.checkpoint(create_custom_forward(resnet), hidden_states, temb)
-                hidden_states = torch.utils.checkpoint.checkpoint(
-                    create_custom_forward(attn, return_dict=False),
-                    hidden_states,
-                    encoder_hidden_states,
-                    cross_attention_kwargs,
-                )[0]
+                if torch.__version__>="1.11.0":
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/versatile_diffusion/modeling_text_unet.py:1431]:
```suggestion
                if is_torch_version(">=", "1.11.0"):
```

**Comment by patrickvonplaten** [comment]
> > Hey @7eu7d7,
> > Could you give a bit more context for this PR?
> 
> When adding adapters (like lora) to the model or training only partial layers of the model (The rest of the parameters do not requires gradients), enable gradient checkpointing can cause some parameters that require gradients cannot be learned. This bug is also existed in the `transformers`.

Cool, ok from my side, also because PyTorch seems to switch to default to `use_reentrant` anyways as stated here: https://pytorch.org/docs/stable/checkpoint.html#torch.utils.checkpoint.checkpoint_sequential

**Comment by patrickvonplaten** [comment]
@7eu7d7,

Could you try to fix the failing tests and also link to the same issue in `transformers` for context?

**Review by patrickvonplaten — APPROVED** [review]
Cool, LGTM! @pcuenca @sayakpaul can you also take a look?

**Comment by 7eu7d7** [comment]
> Could you try to fix the failing tests and also link to the same issue in `transformers` for context?

I open the [same PR in transformer](https://github.com/huggingface/transformers/pull/23422) to fix the bugs and linked to this PR.

**Review by sayakpaul — APPROVED** [review]
LGTM, thank you 🔥 

A dumb question: what happens when PT version is < 1.11.0?
